### ISSUE TITLE
`selectInput` `optgroup` rendering not correct when a group has only one choice

### ISSUE THREAD

**Issue by ksens** [rstudio/shiny#1864]
At https://shiny.rstudio.com/reference/shiny/latest/selectInput.html, you provide an example of grouping similar choices within `selectInput`. 

### Correct rendering when there are multiple choices per group:
The example shows three state choices for each region, and works perfectly.

![screenshot-1](https://user-images.githubusercontent.com/13973052/31078170-1372ca48-a74f-11e7-9fee-14c4667e70f5.png)

### Incorrect rendering when there is one choice in one of the groups
However, when one of the groups has only one choice within it, this is not rendered properly, as is shown in the screenshot below.

![screenshot-2](https://user-images.githubusercontent.com/13973052/31078148-f7facc98-a74e-11e7-9a89-ac5890d2bf8a.png)

### Repro code
```R
## Only run examples in interactive R sessions
if (interactive()) {

# basic example
shinyApp(
  ui = fluidPage(
    selectInput("variable", "Variable:",
                c("Cylinders" = "cyl",
                  "Transmission" = "am",
                  "Gears" = "gear")),
    tableOutput("data")
  ),
  server = function(input, output) {
    output$data <- renderTable({
      mtcars[, c("mpg", input$variable), drop = FALSE]
    }, rownames = TRUE)
  }
)

# demoing optgroup support in the `choices` arg
shinyApp(
  ui = fluidPage(
    selectInput("state", "Choose a state:",
      list(`East Coast` = c("NY", "NJ", "CT"),
           `West Coast` = c("WA", "OR", "CA"),
           `Midwest` = "MN")
    ),
    textOutput("result")
  ),
  server = function(input, output) {
    output$result <- renderText({
      paste("You chose", input$state)
    })
  }
)
}
```

Package description

```R
packageDescription('shiny')
Package: shiny
Type: Package
Title: Web Application Framework for R
Version: 1.0.5
Authors@R: c( person("Winston", "Chang", role = c("aut", "cre"), email = "winston@rstudio.com"), person("Joe",
            "Cheng", role = "aut", email = "joe@rstudio.com"), person("JJ", "Allaire", role = "aut", email =
            "jj@rstudio.com"), person("Yihui", "Xie", role = "aut", email = "yihui@rstudio.com"),
            person("Jonathan", "McPherson", role = "aut", email = "jonathan@rstudio.com"), person(family =
            "RStudio", role = "cph"), person(family = "jQuery Foundation", role = "cph", comment = "jQuery
            library and jQuery UI library"), person(family = "jQuery contributors", role = c("ctb", "cph"),
            comment = "jQuery library; authors listed in inst/www/shared/jquery-AUTHORS.txt"), person(family =
            "jQuery UI contributors", role = c("ctb", "cph"), comment = "jQuery UI library; authors listed in
            inst/www/shared/jqueryui/AUTHORS.txt"), person("Mark", "Otto", role = "ctb", comment = "Bootstrap
            library"), person("Jacob", "Thornton", role = "ctb", comment = "Bootstrap library"), person(family
            = "Bootstrap contributors", role = "ctb", comment = "Bootstrap library"), person(family = "Twitter,
            Inc", role = "cph", comment = "Bootstrap library"), person("Alexander", "Farkas", role = c("ctb",
            "cph"), comment = "html5shiv library"), person("Scott", "Jehl", role = c("ctb", "cph"), comment =
            "Respond.js library"), person("Stefan", "Petre", role = c("ctb", "cph"), comment =
            "Bootstrap-datepicker library"), person("Andrew", "Rowls", role = c("ctb", "cph"), comment =
            "Bootstrap-datepicker library"), person("Dave", "Gandy", role = c("ctb", "cph"), comment =
            "Font-Awesome font"), person("Brian", "Reavis", role = c("ctb", "cph"), comment = "selectize.js
            library"), person("Kristopher Michael", "Kowal", role = c("ctb", "cph"), comment = "es5-shim
            library"), person(family = "es5-shim contributors", role = c("ctb", "cph"), comment = "es5-shim
            library"), person("Denis", "Ineshin", role = c("ctb", "cph"), comment = "ion.rangeSlider library"),
            person("Sami", "Samhuri", role = c("ctb", "cph"), comment = "Javascript strftime library"),
            person(family = "SpryMedia Limited", role = c("ctb", "cph"), comment = "DataTables library"),
            person("John", "Fraser", role = c("ctb", "cph"), comment = "showdown.js library"), person("John",
            "Gruber", role = c("ctb", "cph"), comment = "showdown.js library"), person("Ivan", "Sagalaev", role
            = c("ctb", "cph"), comment = "highlight.js library"), person(family = "R Core Team", role =
            c("ctb", "cph"), comment = "tar implementation from R") )
Description: Makes it incredibly easy to build interactive web applications with R. Automatic "reactive" binding
            between inputs and outputs and extensive prebuilt widgets make it possible to build beautiful,
            responsive, and powerful applications with minimal effort.
License: GPL-3 | file LICENSE
Depends: R (>= 3.0.2), methods
Imports: utils, httpuv (>= 1.3.5), mime (>= 0.3), jsonlite (>= 0.9.16), xtable, digest, htmltools (>= 0.3.5), R6
            (>= 2.0), sourcetools, tools
Suggests: datasets, Cairo (>= 1.5-5), testthat, knitr (>= 1.6), markdown, rmarkdown, ggplot2, magrittr
URL: http://shiny.rstudio.com
BugReports: https://github.com/rstudio/shiny/issues
Collate: 'app.R' 'bookmark-state-local.R' 'stack.R' 'bookmark-state.R' 'bootstrap-layout.R' 'conditions.R'
            'map.R' .....
RoxygenNote: 6.0.1
NeedsCompilation: no
Packaged: 2017-08-23 18:15:30 UTC; winston
Author: Winston Chang [aut, cre], Joe Cheng [aut], JJ Allaire [aut], Yihui Xie [aut], Jonathan McPherson [aut],
            RStudio [cph], jQuery Foundation [cph] (jQuery library and jQuery UI library), jQuery contributors
            [ctb, cph] (jQuery library; authors listed in inst/www/shared/jquery-AUTHORS.txt), jQuery UI
            contributors [ctb, cph] (jQuery UI library; authors listed in
            inst/www/shared/jqueryui/AUTHORS.txt), Mark Otto [ctb] (Bootstrap library), Jacob Thornton [ctb]
            (Bootstrap library), Bootstrap contributors [ctb] (Bootstrap library), Twitter, Inc [cph]
            (Bootstrap library), Alexander Farkas [ctb, cph] (html5shiv library), Scott Jehl [ctb, cph]
            (Respond.js library), Stefan Petre [ctb, cph] (Bootstrap-datepicker library), Andrew Rowls [ctb,
            cph] (Bootstrap-datepicker library), Dave Gandy [ctb, cph] (Font-Awesome font), Brian Reavis [ctb,
            cph] (selectize.js library), Kristopher Michael Kowal [ctb, cph] (es5-shim library), es5-shim
            contributors [ctb, cph] (es5-shim library), Denis Ineshin [ctb, cph] (ion.rangeSlider library),
            Sami Samhuri [ctb, cph] (Javascript strftime library), SpryMedia Limited [ctb, cph] (DataTables
            library), John Fraser [ctb, cph] (showdown.js library), John Gruber [ctb, cph] (showdown.js
            library), Ivan Sagalaev [ctb, cph] (highlight.js library), R Core Team [ctb, cph] (tar
            implementation from R)
Maintainer: Winston Chang <winston@rstudio.com>
Repository: CRAN
Date/Publication: 2017-08-23 20:20:10 UTC
Built: R 3.4.0; ; 2017-09-05 03:21:42 UTC; unix
RemoteType: cran
RemoteSha: 1.0.5
RemoteRepos: structure("https://cloud.r-project.org", .Names = "CRAN")
RemotePkgType: source

-- File: /usr/lib64/R/library/shiny/Meta/package.rds
```

**Comment by bborgesr** [comment]
I can repro. @rstudio/shiny -- I think we can take care of this for the next release -- it should be a relatively minor fix... Let me know if there's any objections.

**Comment by ksens** [comment]
Great. Thanks!

**Comment by octaviancorlade** [comment]
From the examples and the tests it looks like it is made this way to allow the first options in the list to be unlabeled, keeping the sublists at the end, such as:

```R
shinyApp(
    ui = fluidPage(
        selectInput("state", "Choose a state:",
                    list("Unknown coast 1", "Unknown coast 2",
                         `East Coast` = c("NY", "NJ", "CT"),
                         `West Coast` = c("WA", "OR", "CA")
                        )
        )
    ), server = function(input, output) {}
)
```

<img width="309" alt="screen shot 2017-10-02 at 17 53 24" src="https://user-images.githubusercontent.com/15089539/31086640-a66d5f84-a79a-11e7-8bff-2a409a714f16.png">

Using lists currently works as a work-around to display single-element sublists:
```R
shinyApp(
    ui = fluidPage(
        selectInput("state", "Choose a state:",
                    list(`East Coast` = list("NY", "NJ", "CT"),
                         `West Coast` = list("WA", "OR", "CA"),
                         `Midwest` = list("MN")
                        )
        )
    ), server = function(input, output) {}
)
```
<img width="310" alt="screen shot 2017-10-02 at 17 57 07" src="https://user-images.githubusercontent.com/15089539/31086796-2402f382-a79b-11e7-8f01-19f8b3a3b21b.png">

**Comment by wch** [comment]
I think the current behavior is correct. When running the app with this:

```R
    selectInput("state", "Choose a state:",
      list(`East Coast` = c("NY", "NJ", "CT"),
           `West Coast` = c("WA", "OR", "CA"),
           `Midwest` = "MN")
    )
```

The Midwest option is not in the West Coast group; it is an option at the top level. However, Selectize renders it in a way where it doesn't look like it's separate from the previous group.

![image](https://user-images.githubusercontent.com/86978/31089287-1d487d00-a769-11e7-9a03-bbdf108ae0d3.png)

The documentation could be improved, though. Using a list instead of a char vector makes the grouping unambiguous.

Another alternative is to use a named vector. This will put the MN option in the Midwest group:

```R
    selectInput("state", "Choose a state:",
      list(`East Coast` = c("NY", "NJ", "CT"),
           `West Coast` = c("WA", "OR", "CA"),
           `Midwest` = c("MN" = "MN"))
    )
```

**Comment by bborgesr** [comment]
@wch, I see. The reason this struck me as odd is because of this:

```r
    selectInput("state", "Choose a state:",
      list(`East Coast` = c("NY", "NJ", "CT"),
           `West Coast` = c("WA", "OR", "CA"),
           `Midwest` = c("MN"))
    )
```

It is unintuitive for people to figure out they need to use `list()`/named vector when an unnamed vector works fine for the previous cases. 

I guess the best thing (including for backward compatibility) would be to just change documentation/gallery examples to always use `list()` when `optgoup`ing.

What do you think?

**Comment by wch** [comment]
@bborgesr I agree, using `list()` in the examples is probably the best thing to do.

**Comment by ksens** [comment]
Substituting with `list()` worked perfectly for me.

**Comment by kjbennet** [comment]
I have also encountered this problem when I populate selectInput using a SQL query, which I think results in an automatic optgroup format. Unfortunately I can't figure out how to use the list() workaround in this case.

**Comment by wch** [comment]
Closed by #2084.
